src/builtin.cpp

```diff
@@ -2158,104 +2158,218 @@ static int builtin_read(parser_t &parser, io_streams_t &streams, wchar_t **argv)
-        IS_LOGIN,
-        IS_FULL_JOB_CONTROL,
-        IS_INTERACTIVE_JOB_CONTROL,
-        IS_NO_JOB_CONTROL,
-        STACK_TRACE,
-        DONE,
-        CURRENT_FILENAME,
-        CURRENT_LINE_NUMBER
-    };
+enum status_cmd_t {
+    STATUS_NOOP,
+    STATUS_IS_LOGIN,
+    STATUS_IS_INTERACTIVE,
+    STATUS_IS_BLOCK,
+    STATUS_IS_COMMAND_SUB,
+    STATUS_IS_FULL_JOB_CTRL,
+    STATUS_IS_INTERACTIVE_JOB_CTRL,
+    STATUS_IS_NO_JOB_CTRL,
+    STATUS_CURRENT_FILENAME,
+    STATUS_CURRENT_LINE_NUMBER,
+    STATUS_SET_JOB_CONTROL,
+    STATUS_PRINT_STACK_TRACE
+};
+
+static status_cmd_t status_string_to_cmd(const wchar_t *status_command) {
+    if (wcscmp(status_command, L"is-login") == 0) {
+        return STATUS_IS_LOGIN;
+    } else if (wcscmp(status_command, L"is-interactive") == 0) {
+        return STATUS_IS_INTERACTIVE;
+    } else if (wcscmp(status_command, L"is-block") == 0) {
+        return STATUS_IS_BLOCK;
+    } else if (wcscmp(status_command, L"is-command-sub") == 0) {
+        return STATUS_IS_COMMAND_SUB;
+    } else if (wcscmp(status_command, L"is-full-job-control") == 0) {
+        return STATUS_IS_FULL_JOB_CTRL;
+    } else if (wcscmp(status_command, L"is-interactive-job-control") == 0) {
+        return STATUS_IS_NO_JOB_CTRL;
+    } else if (wcscmp(status_command, L"is-no-job-control") == 0) {
+        return STATUS_IS_INTERACTIVE_JOB_CTRL;
+    } else if (wcscmp(status_command, L"current-filename") == 0) {
+        return STATUS_CURRENT_FILENAME;
+    } else if (wcscmp(status_command, L"current-line-number") == 0) {
+        return STATUS_CURRENT_LINE_NUMBER;
+    } else if (wcscmp(status_command, L"job-control") == 0) {
+        return STATUS_SET_JOB_CONTROL;
+    } else if (wcscmp(status_command, L"print-stack-trace") == 0) {
+        return STATUS_PRINT_STACK_TRACE;
+    }
+    return STATUS_NOOP;
+}
 
-    int mode = NORMAL;
-    int argc = builtin_count_args(argv);
-    int res = STATUS_BUILTIN_OK;
+static const wcstring status_cmd_to_string(status_cmd_t status_cmd) {
+    switch (status_cmd) {
+        case STATUS_NOOP:
+            return L"no-op";
+        case STATUS_IS_LOGIN:
+            return L"is-login";
+        case STATUS_IS_INTERACTIVE:
+            return L"is-interactive";
+        case STATUS_IS_BLOCK:
+            return L"is-block";
+        case STATUS_IS_COMMAND_SUB:
+            return L"is-command-sub";
+        case STATUS_IS_FULL_JOB_CTRL:
+            return L"is-full-job-control";
+        case STATUS_IS_INTERACTIVE_JOB_CTRL:
+            return L"is-interactive-job-control";
+        case STATUS_IS_NO_JOB_CTRL:
+            return L"is-no-job-control";
+        case STATUS_CURRENT_FILENAME:
+            return L"current-filename";
+        case STATUS_CURRENT_LINE_NUMBER:
+            return L"current-line-number";
+        case STATUS_SET_JOB_CONTROL:
+            return L"job-control";
+        case STATUS_PRINT_STACK_TRACE:
+            return L"print-stack-trace";
+    }
+}
```

Comment:
I'm just about to create a PR that generalizes that code using table driven template methods. That should take care of those false positives.